AI/src/search/mcts/nodes/PNMCTSNode.java

```diff
@@ -0,0 +1,339 @@
+	
+	//-------------------------------------------------------------------------
+    
+    /**
+     * Constructor 
+     * 
+     * @param mcts
+     * @param parent
+     * @param parentMove
+     * @param parentMoveWithoutConseq
+     * @param context
+     */
+    public PNMCTSNode
+    (
+    	final MCTS mcts, 
+    	final BaseNode parent, 
+    	final Move parentMove, 
+    	final Move parentMoveWithoutConseq,
+    	final Context context
+    )
+    {
+    	super(mcts, parent, parentMove, parentMoveWithoutConseq, context);
+    	
+    	if (parent == null)
+    	{
+    		// We are the root node
+    		rootPlayer = context.state().mover();
+    	}
+    	else
+    	{
+    		rootPlayer = ((PNMCTSNode) parent).rootPlayer;
+    	}
+    	
+    	if (context.state().mover() == rootPlayer) 
+    	{
+            this.type = PNMCTSNodeTypes.OR_NODE;
+        }
+    	else 
+    	{
+            this.type = PNMCTSNodeTypes.AND_NODE;
+        }
+    	
+    	evaluate();
+        setProofAndDisproofNumbers();
+        
+        childrenPNSSelectionTerms = new double[numLegalMoves()];
+    }
+    
+    //-------------------------------------------------------------------------
+    
+    /**
+     * Evaluates a node as in PNS according to L. V. Allis' 
+     * "Searching for Solutions in Games and Artificial Intelligence"
+     */
+    public void evaluate() 
+    {
+        if (this.context.trial().over()) 
+        {
+            if (RankUtils.utilities(this.context)[rootPlayer] == 1.0) 
+            {
+                this.proofValue = PNMCTSNodeValues.TRUE;
+            } 
+            else 
+            {
+                this.proofValue = PNMCTSNodeValues.FALSE;
+            }
+        } 
+        else 
+        {
+            this.proofValue = PNMCTSNodeValues.UNKNOWN;
+        }
+    }
+    
+    /**
+     * Sets the proof and disproof values of the current node as it is done for 
+     * PNS in L. V. Allis' "Searching for Solutions in Games and Artificial 
+     * Intelligence". Set differently depending on if the node has children yet.
+     *
+     * @return Returns true if something was changed and false if not. 
+     * Used to improve PN-MCTS speed.
+     */
+    public boolean setProofAndDisproofNumbers() 
+    {
+        if (legalMoves.length > 0) 
+        {
+        	// Not a terminal node
+        	double proof;
+        	double disproof;
+        	switch (type)
+        	{
+			case AND_NODE:
+				proof = 0.0;
+				disproof = Double.POSITIVE_INFINITY;
+				for (final BaseNode child : children)
+				{
+					final PNMCTSNode childNode = (PNMCTSNode) child;
+					if (childNode != null)
+					{
+						proof += childNode.proofNumber;
+						
+						if (childNode.disproofNumber < disproof)
+						{
+							disproof = childNode.disproofNumber;
+						}
+					}
+					else
+					{
+						// An unexpanded child
+						// TODO verify that this is indeed what we want to do?
+						// NOTE: addition for proof, setting for disproof
```

Comment:
Jakub please check if this is what we want for children that have not been added to tree yet